### DISCUSSION TITLE
Get pos_ydir coordinates from a pre-defined position (coil center) and angle

### DISCUSSION THREAD

**Discussion by billotanne** [simnibs/simnibs#321] [8. Simulation workflows] [answered]
Dear SimNIBS Experts,

We would like to run simnibs to simulate the TMS E-field at a specific position (x, y, z coordinates) using a specific coil angle (45 degrees). How can we convert an angle (e.g., 45 degrees) at a specific position to orientation/direction coordinates (`pos_ydir`) used in simnibs when running the E-field simulation? Is there another way to directly specify the angle in `s.poslist` instead of specifying the direction coordinates (or the 4x4 matsimnibs matrix)?

Thank you very much,
Anne

**Comment by oulap** [comment, marked as answer]
Hi Anne,

You can multiply the simnibs coil matrix with a rotation matrix constructed to rotate around the z-axis. There's an example of how to construct such a matrix (Rz) here:

https://github.com/simnibs/simnibs/blob/d045e8ccda6b349cbec639563af32dac69e95909/simnibs/optimization/optimize_tms.py#L66

The input to the function (R) would be the simnibs coil matrix. If you define the  rotation angle to be 45 degrees, in the end you'll get a new matrix where the ydir is rotated 45 degrees from where it was pointing before.

**Reply by GuillaumeCDGJ** [reply]
Hello,

I am trying to perform a similar simulation, in which I only have the x, y, z coordinates of my target and the angle of the coil handle relative to the horizontal plane at my disposal. From what I understand, this simnibs coil matrix you refer to is usually imported from whichever neuronavigation system was used. However, I don't have access to those files. Is there any way to circumvent this issue, knowing that the coils were placed tangentially to the scalp surface, with only the cortical target coordinates and coil orientation ?

Thank you in advance,

Guillaume

**Reply by oulap** [reply]
Hi Guillaume,

You can see how the coil matrix is constructed here:

https://github.com/simnibs/simnibs/blob/d045e8ccda6b349cbec639563af32dac69e95909/simnibs/mesh_tools/mesh_io.py#L1866

Basically you need the center and some y-direction that you can define yourself. The coil is by default placed tangentially to the scalp surface (by taking the normals of the scalp elements). For your purposes I think you just need to be able to define the y-direction as a point relative to the center of the coil from the the angle and plane you have.

**Reply by GuillaumeCDGJ** [reply]
Thank you for your quick reply.

However, after having generated a coil matrix with the handle on the horizontal plane using ```calc_matsimnibs``` , it appears it is not possible to use ```_rotate system``` on it since the latter expects a 3D matrix and the former returns a 4D one. Am I going about it wrong ?

Best,

Guillaume

**Reply by oulap** [reply]
The rotation only  considers the 3x3 part of the transformation matrix (leaving out the translation), so you can try to pass only that part to the _rotate_system function, i.e., something like:
`
_rotate_system(mat[:3,:3], angle_limits, angle_res)`

where mat is the 4x4 matrix you created using calc_matsimnibs

**Reply by GuillaumeCDGJ** [reply]
Thank you very much, I now have the rotated 3x3 matrix. If I understand correctly, the last step before running the simulation is to concatenate it with the translation column and assign the resulting 4x4 matrix to ```pos.matsimnibs``` with pos being a POSITION variable ?

**Reply by oulap** [reply]
Yeah, you can just assign the rotated matrix back to the original one, i.e., something like:

mat[:3, :3] = rotated

and then use that matrix in the simulation as you mentioned.

**Reply by GuillaumeCDGJ** [reply]
Thank you very much, you've been really helpful.

Best,

Guillaume

**Reply by oulap** [reply]
No problem!

**Reply by billotanne** [reply]
Hello, 

I’m following up on my initial question, how would you get an initial simnibs coil matrix at 0 degrees to the sagittal plane if you have xyz coordinates of the coil center? (to then rotate it to 45 degrees) In other words, how do you get pos_ydir for 0 degrees?

Thanks,
Anne

**Reply by oulap** [reply]
Hi Anne,

Have a look at this code snippet, which constructs the matsimnibs:

https://github.com/simnibs/simnibs/blob/d045e8ccda6b349cbec639563af32dac69e95909/simnibs/mesh_tools/mesh_io.py#L1901

The y-direction is defined from the center of the coil and pos_ydir. If you want the coil pointing towards the anterior direction you can try setting the pos_ydir like this:

pos_ydir = center + 20*[0, 1, 0]

i.e., add 20 mm to the y-coordinate of the center. You can then call the calc_matsimnibs function (linked above) to get the initial matrix, which you can then rotate. Hope that makes sense.

**Reply by billotanne** [reply]
Makes sense! Thanks so much for the quick reply!

Anne

**Comment by billotanne** [comment]
Thanks very much for the suggestion!
We'll try that!

Best,

Anne
